File: mods/bds/src/state/bvcs_codec.ts
// Decoder for the no-net reverse-ride grammar the desktop proxy injects as
// serverbound chat. Mirrors common/src/structs/control/bvcs_codec.rs — the
// common-side goldens pin the strings; keep both ends in sync:
//
//   !bvcs:<seq>:q:m=<0|1>;d=<0|1>;r=<0|1>;g=<group|->   self-state snapshot
//   !bvcs:<seq>:p:t=<target>;v=<percent>;h=<0|1>        one player preference
//   !bvcs:<seq>:gl:c=<count>                            group-list header
//   !bvcs:<seq>:g:i=<id>;n=<name>                       one group

export interface BvcsQueryState {
  kind: 'q';
  muted: boolean;
  deafened: boolean;
  recording: boolean;
  group: string | null;
}

export interface BvcsPreference {
  kind: 'p';
  target: string;
  volume: number;
  heard: boolean;
}

export interface BvcsGroupListHeader {
  kind: 'gl';
  count: number;
}

export interface BvcsGroup {
  kind: 'g';
  id: string;
  name: string;
}

export type BvcsMessage =
  | BvcsQueryState
  | BvcsPreference
  | BvcsGroupListHeader
  | BvcsGroup;

const BVCS_PREFIX = '!bvcs:';

export class BvcsCodec {
  static isBvcs(message: string): boolean {
    return message.startsWith(BVCS_PREFIX);
  }

  static decode(message: string): BvcsMessage | null {
    if (!message.startsWith(BVCS_PREFIX)) {
      return null;
    }
    const rest = message.slice(BVCS_PREFIX.length);
    const first = rest.indexOf(':');
    if (first < 0) {
      return null;
    }
    const second = rest.indexOf(':', first + 1);
    if (second < 0) {
      return null;
    }
    if (!/^\d+$/.test(rest.slice(0, first))) {
      return null;
    }

    const kind = rest.slice(first + 1, second);
    const fields = new Map<string, string>();
    for (const kv of rest.slice(second + 1).split(';')) {
      const eq = kv.indexOf('=');
      if (eq > 0) {
        fields.set(kv.slice(0, eq), kv.slice(eq + 1));
      }
    }

    if (kind === 'q') {
      const m = fields.get('m');
      const d = fields.get('d');
      const r = fields.get('r');
      const g = fields.get('g');
      if (m === undefined || d === undefined || r === undefined) {
        return null;
      }
      return {
        kind: 'q',
        muted: m === '1',
        deafened: d === '1',
        recording: r === '1',
        group: g !== undefined && g !== '-' ? g : null,
      };
    }

    if (kind === 'p') {
      const t = fields.get('t');
      const v = fields.get('v');
      const h = fields.get('h');
      if (!t || v === undefined || h === undefined) {
        return null;
      }
      const percent = Number(v);
      if (!Number.isFinite(percent)) {
        return null;
      }
      return {
        kind: 'p',
        target: t,
        volume: percent / 100,
        heard: h === '1',
      };
    }

    if (kind === 'gl') {
      const c = fields.get('c');
      if (c === undefined || !/^\d+$/.test(c)) {
        return null;
      }
      return { kind: 'gl', count: Number(c) };
    }

    if (kind === 'g') {
      const id = fields.get('i');
      const encoded = fields.get('n');
      if (id === undefined || encoded === undefined) {
        return null;
      }
      const name = BvcsCodec.unescape(encoded);
      // A name that cannot be decoded means a corrupt ride; a row with a mangled
      // name is worse than no row.
      return name === null ? null : { kind: 'g', id, name };
    }

    return null;
  }

  // Mirrors BvcsCodec::unescape in common. The grammar delimits on `;`, `=` and
  // `:`, which a renamed group may contain.
  private static unescape(value: string): string | null {
    let out = '';
    for (let i = 0; i < value.length; i++) {
      if (value[i] !== '%') {
        out += value[i];
        continue;
      }
      const hex = value.slice(i + 1, i + 3);
      if (!/^[0-9a-fA-F]{2}$/.test(hex)) {
        return null;
      }
      out += String.fromCharCode(parseInt(hex, 16));
      i += 2;
    }
    return out;
  }
}

File: common/src/structs/control/bvcs_codec.rs
use super::{PlayerPreference, QueryState};

pub const BVCS_PREFIX: &str = "!bvcs:";

// Decoded `!bvcs:` reverse-ride message. The wire omits the player identity —
// the BDS listener attributes each message to the chat sender, exactly as the
// proxy injected it on that player's own session.
#[derive(Debug, Clone, PartialEq)]
pub enum BvcsMessage {
    QueryState {
        muted: bool,
        deafened: bool,
        recording: bool,
        group: Option<String>,
    },
    Preference {
        target: String,
        volume: f32,
        muted: bool,
    },
    GroupListHeader {
        count: usize,
    },
    Group {
        id: String,
        name: String,
    },
}

/// Encodes/decodes the no-net reverse-ride grammar the proxy injects as
/// serverbound chat and the standalone BDS mod parses
/// (mods/bds/src/state/bvcs_codec.ts — keep both ends in sync; the common-side
/// goldens pin the strings):
///
/// `!bvcs:<seq>:q:m=<0|1>;d=<0|1>;r=<0|1>;g=<group|->` — self-state snapshot
/// `!bvcs:<seq>:p:t=<target>;v=<percent>;h=<0|1>`      — one player preference
///
/// Volume rides as a rounded percent and mute as an inverted heard flag,
/// mirroring the `bvc:ctl:` grammar's conventions. `<seq>` is a monotonic
/// message tag; each message is bounded far below the TextPacket limit, so no
/// chunking path exists.
pub struct BvcsCodec;

impl BvcsCodec {
    pub fn encode_query_state(seq: u64, state: &QueryState) -> String {
        format!(
            "{BVCS_PREFIX}{seq}:q:m={};d={};r={};g={}",
            state.muted as u8,
            state.deafened as u8,
            state.recording as u8,
            state.current_group.as_deref().unwrap_or("-"),
        )
    }

    /// A target is wire-safe when it cannot corrupt the `key=value;` grammar.
    /// Xbox gamertags never contain these characters; a future identity source
    /// that does must not ride until the grammar gains escaping.
    pub fn target_is_wire_safe(target: &str) -> bool {
        !target.contains([';', '=', ':'])
    }

    pub fn encode_preference(seq: u64, preference: &PlayerPreference) -> String {
        format!(
            "{BVCS_PREFIX}{seq}:p:t={};v={};h={}",
            preference.target,
            (preference.volume * 100.0).round() as u32,
            if preference.muted { 0 } else { 1 },
        )
    }

    /// Announces how many `g` rows follow. A receiver clears its list on this message,
    /// so rows from an earlier request that arrive afterwards are discarded rather than
    /// mixed into the new one.
    pub fn encode_group_header(seq: u64, count: usize) -> String {
        format!("{BVCS_PREFIX}{seq}:gl:c={count}")
    }

    pub fn encode_group(seq: u64, id: &str, name: &str) -> String {
        format!("{BVCS_PREFIX}{seq}:g:i={id};n={}", Self::escape(name))
    }

    pub fn decode(message: &str) -> Option<BvcsMessage> {
        let rest = message.strip_prefix(BVCS_PREFIX)?;
        let mut it = rest.splitn(3, ':');
        let _seq: u64 = it.next()?.parse().ok()?;
        let kind = it.next()?;
        let fields = Self::parse_fields(it.next()?);

        match kind {
            "q" => Some(BvcsMessage::QueryState {
                muted: fields.iter().find(|(k, _)| *k == "m")?.1 == "1",
                deafened: fields.iter().find(|(k, _)| *k == "d")?.1 == "1",
                recording: fields.iter().find(|(k, _)| *k == "r")?.1 == "1",
                group: fields
                    .iter()
                    .find(|(k, _)| *k == "g")
                    .and_then(|(_, v)| (*v != "-").then(|| v.to_string())),
            }),
            "p" => {
                let target = fields.iter().find(|(k, _)| *k == "t")?.1.to_string();
                let percent: f32 = fields.iter().find(|(k, _)| *k == "v")?.1.parse().ok()?;
                if !percent.is_finite() {
                    return None;
                }
                let heard = fields.iter().find(|(k, _)| *k == "h")?.1 == "1";
                Some(BvcsMessage::Preference {
                    target,
                    volume: percent / 100.0,
                    muted: !heard,
                })
            }
            "gl" => {
                let count: usize = fields.iter().find(|(k, _)| *k == "c")?.1.parse().ok()?;
                Some(BvcsMessage::GroupListHeader { count })
            }
            "g" => {
                let id = fields.iter().find(|(k, _)| *k == "i")?.1.to_string();
                let name = Self::unescape(fields.iter().find(|(k, _)| *k == "n")?.1)?;
                Some(BvcsMessage::Group { id, name })
            }
            _ => None,
        }
    }

    fn parse_fields(payload: &str) -> Vec<(&str, &str)> {
        payload
            .split(';')
            .filter_map(|kv| kv.split_once('='))
            .collect()
    }

    // The grammar delimits on `;`, `=` and `:`, which a renamed group may contain.
    // `%` is encoded first so decoding is unambiguous.
    fn escape(value: &str) -> String {
        let mut out = String::with_capacity(value.len());
        for c in value.chars() {
            match c {
                '%' => out.push_str("%25"),
                ';' => out.push_str("%3B"),
                '=' => out.push_str("%3D"),
                ':' => out.push_str("%3A"),
                _ => out.push(c),
            }
        }
        out
    }

    fn unescape(value: &str) -> Option<String> {
        let bytes = value.as_bytes();
        let mut out = Vec::with_capacity(bytes.len());
        let mut index = 0;
        while index < bytes.len() {
            if bytes[index] == b'%' {
                let hex = value.get(index + 1..index + 3)?;
                out.push(u8::from_str_radix(hex, 16).ok()?);
                index += 3;
            } else {
                out.push(bytes[index]);
                index += 1;
            }
        }
        String::from_utf8(out).ok()
    }
}
